### PR TITLE
`EnvironmentType` design tweaks

### PR DESCRIPTION
**Pull request by serhii-lekariev**
This PR simplifies `EnvironmentType` and `Environment`.

#### `Environment`

Now, allows to check for the class with `boolean is(Class<? extends EnvironmentType>), instead of `boolean is(EnvironmentType)`.

#### `EnvironmentType` 

Now, no longer overrides `equals` and `hashCode` in suspicious ways. Instead, it simply acts as a stricter `interface` for a single `boolean enabled()` method.

Also, standard environment types are no longer singletons. We no longer encourage the extenders to be singletons as well.

Also resolves some Error Prone warnings.

### CODE REVIEW

**Comment by serhii-lekariev** [comment]
@dmdashenkov, PTAL.

File: base/src/main/java/io/spine/base/Environment.java [hunk 1]
```diff
@@ -78,19 +77,19 @@
  *
  *     static {
  *         Environment.instance()
- *                    .register(StagingEnvironmentType.type())
- *                    .register(LoadTestingType.type());
+ *                    .register(new StagingEnvironmentType())
```
Comment (dmdashenkov) [inline base/src/main/java/io/spine/base/Environment.java:80]: Let's skip `Type` here, just like we don't call it `TestsType` and `ProductionType`.

File: base/src/main/java/io/spine/base/Environment.java [hunk 2]
```diff
@@ -181,50 +180,22 @@ public Environment createCopy() {
...
-    private void determineCurrentType() {
-        for (EnvironmentType type : knownEnvTypes) {
-            if (type.enabled()) {
-                this.currentEnvType = type;
-                return;
-            }
-        }
+    public boolean is(Class<? extends EnvironmentType> typeClass) {
+        EnvironmentType currentType = cachedOrCalculated();
+        boolean result = currentType.getClass()
```
Comment (dmdashenkov) [inline base/src/main/java/io/spine/base/Environment.java:185]: I'd go with `type.isInstance(..)`. That will give us an opportunity to create polymorphic types.

File: base/src/main/java/io/spine/base/Environment.java [hunk 3]
```diff
@@ -181,50 +180,22 @@ public Environment createCopy() {
...
-    }
-
-    private void determineCurrentType() {
-        for (EnvironmentType type : knownEnvTypes) {
-            if (type.enabled()) {
-                this.currentEnvType = type;
-                return;
-            }
-        }
+    public boolean is(Class<? extends EnvironmentType> typeClass) {
```
Comment (dmdashenkov) [inline base/src/main/java/io/spine/base/Environment.java:183]:
```suggestion
    public boolean is(Class<? extends EnvironmentType> type) {
```

File: base/src/main/java/io/spine/base/Environment.java [hunk 4]
```diff
@@ -298,4 +269,22 @@ public void reset() {
...
+
+    private EnvironmentType currentType() {
+        for (EnvironmentType type : knownEnvTypes) {
+            if (type.enabled()) {
+                return type;
+            }
+        }
+
+        // `Production` is the default fallback.
+        return new Production();
```
Comment (dmdashenkov) [inline base/src/main/java/io/spine/base/Environment.java:288]: Seems like an illegal state to me.

**Comment by serhii-lekariev** [comment]
@dmdashenkov, PTAL again.

**Pull request merged**